### PR TITLE
Add ability to dump ticket.db and decrypt titlekeys directly from NAND

### PR DESCRIPTION
**Pull request by d0k3**
As always, if anything should be changed just let me know and I'll make the required changes.

### CODE REVIEW

File: source/decryptor/decryptor.c [hunk 1]
```diff
@@ -113,7 +132,74 @@ u32 DecryptTitlekeys(void)
     }
     FileClose();
 
-    Debug("Done!");
```
Comment (d0k3) [inline source/decryptor/decryptor.c:116]: This message is redundant. The success message is given after completion of each feature function anyways.

File: source/decryptor/decryptor.c [hunk 2]
```diff
@@ -113,7 +132,74 @@ u32 DecryptTitlekeys(void)
...
+{
+    PartitionInfo* ctrnand_info = &(partitions[(GetUnitPlatform() == PLATFORM_3DS) ? 5 : 6]);
+    u8* buffer = BUFFER_ADDRESS;
+    EncKeysInfo *info = (EncKeysInfo*) 0x20316000;
+    u32 nKeys = 0;
+    u8* titlekey;
+    u8* titleId;
+    u32 commonKeyIndex;
+    u32 offset;
+    u32 size;
```
Comment (archshift) [inline source/decryptor/decryptor.c]: This isn't C89, you don't have to declare all variables at the top of their scope. It's much better to declare the variables closest to where you're going to use them.

File: source/decryptor/decryptor.c [hunk 3]
```diff
@@ -348,6 +434,64 @@ u32 GetNandCtr(u8* ctr, u32 offset)
...
+    
+    const static char* magic = ".          ";
+    const static char zeroes[8+3] = { 0x00 };
+    u8* buffer = BUFFER_ADDRESS;
+    u32 p_size = partition->size;
+    u32 p_offset = partition->offset;
+    
+    u32 cluster_size;
+    u32 cluster_start;
+    u32 found = 0;
```
Comment (archshift) [inline source/decryptor/decryptor.c]: Use `bool` instead.

File: source/decryptor/decryptor.c [hunk 4]
```diff
@@ -348,6 +434,64 @@ u32 GetNandCtr(u8* ctr, u32 offset)
...
+        NAND_SECTOR_SIZE * ( *((u16*) (buffer + 0x0E)) + // FAT table start
+        ( *((u16*) (buffer + 0x16)) * buffer[0x10] ) ) + // FAT table size
+        *((u16*) (buffer + 0x11)) * 0x20; // root directory size
+    cluster_size = buffer[0x0D] * NAND_SECTOR_SIZE;
+    if (seekpos != NULL) {
+        if (cluster_start > *seekpos)
+            *seekpos = cluster_start;
+    }
+    
+    for( u32 i = (seekpos == NULL) ? cluster_start : *seekpos; i < p_size; i += cluster_size ) {
```
Comment (archshift) [inline source/decryptor/decryptor.c]: Spacing:

```
for (u32 i = (seekpos == NULL) ? cluster_start : *seekpos; i < p_size; i += cluster_size)
```

File: source/decryptor/decryptor.c [hunk 5]
```diff
@@ -348,6 +434,64 @@ u32 GetNandCtr(u8* ctr, u32 offset)
...
+    u8* buffer = BUFFER_ADDRESS;
+    u32 p_size = partition->size;
+    u32 p_offset = partition->offset;
+    
+    u32 cluster_size;
+    u32 cluster_start;
+    u32 found = 0;
+    
+    
+    if (strlen(filename) != 8+3)
```
Comment (archshift) [inline source/decryptor/decryptor.c]: I think we can use `strnlen` with devkitARM, if so you should use that instead.

File: source/decryptor/decryptor.c [hunk 6]
```diff
@@ -348,6 +434,64 @@ u32 GetNandCtr(u8* ctr, u32 offset)
...
+        return 1;
+    
+    DecryptNandToMem(buffer, p_offset, NAND_SECTOR_SIZE, partition);
+    cluster_start =
+        NAND_SECTOR_SIZE * ( *((u16*) (buffer + 0x0E)) + // FAT table start
```
Comment (archshift) [inline source/decryptor/decryptor.c]: This is quite confusing to read, can you split these out into separate variables instead?

File: source/decryptor/decryptor.c [hunk 7]
```diff
@@ -348,6 +434,64 @@ u32 GetNandCtr(u8* ctr, u32 offset)
...
+        ( *((u16*) (buffer + 0x16)) * buffer[0x10] ) ) + // FAT table size
+        *((u16*) (buffer + 0x11)) * 0x20; // root directory size
+    cluster_size = buffer[0x0D] * NAND_SECTOR_SIZE;
+    if (seekpos != NULL) {
+        if (cluster_start > *seekpos)
```
Comment (archshift) [inline source/decryptor/decryptor.c]:
``` cpp
if (seekpos != NULL && cluster_start > *seekpos)
```

File: source/decryptor/decryptor.c [hunk 8]
```diff
@@ -348,6 +434,64 @@ u32 GetNandCtr(u8* ctr, u32 offset)
...
+            *seekpos = cluster_start;
+    }
+    
+    for( u32 i = (seekpos == NULL) ? cluster_start : *seekpos; i < p_size; i += cluster_size ) {
+        DecryptNandToMem(buffer, p_offset + i, NAND_SECTOR_SIZE, partition);
+        if (memcmp(buffer, magic, 8+3) == 0) {
```
Comment (archshift) [inline source/decryptor/decryptor.c]: Can you reduce the level of indentations in this function by `continue`ing instead of going down the `if` chain?

**Comment by d0k3** [comment]
Alright, I did all modifications you suggested, compiled and tested it and squashed them into the previous commits..

**Comment by d0k3** [comment]
Alright, there are some features currently in my wip version, that I would like to ask you about before doing a pull request for them:

**Using a working directory**
Use a work directory instead of the SD card root for all operations. That means all input (seeddb.bin, ncchinfo.bin, ...) and output (twln.bin, dectitlekeys.bin, ...) go there. In theory that's a good thing, because it prevents cluttering of the SD card. In daily use, however, it leads to tons of noob questions, because basically all tutorials say files have to be put in the SD card root. It is disabled for now in my wip version. Still unsure about that one, though.

**(1) NAND restore and (2) NAND partition injection**
This restores the full NAND from a backup (1) or encrypts and injects certain NAND partitions (2). It has its 'mainstream' uses, like f.e. editing the TWLN partition to run DSIware games. It is also, when used incorrectly, dangerous and can easily lead to a brick. In my wip version this has to be enabled via a compiler switch and is disabled by default. These features have been around for some time in my wip version now and are tested by multiple users. Unsure if you would want something like this in the Decrypt9 main version.

**Operations on EmuNAND**
Pretty easy to implement and also not dangerous. Dumping the ticket.db from EmuNAND is an often requested feature, other candidates would by decrypting titlekeys from EmuNAND, EmuNAND backup and dumping EmuNAND partitions. Everything besides EmuNAND backup works universally and is not limited to specific kinds of EmuNAND (like, f.e. the Gateway EmuNAND).

Let me know what you think about these.

**Comment by archshift** [comment]
Those last two sound pretty good to me. Like you mentioned, the first one will probably cause lots of confusion so for now I'm a little hesitant on that.

**Pull request merged**